tokio-tasks/src/task_map.rs

```diff
@@ -0,0 +1,70 @@
+use crate::future_ext::FutureExt;
+use futures::future::RemoteHandle;
+use futures::Future;
+use std::collections::HashMap;
+use std::hash;
+
+/// Struct controlling the lifetime of the async tasks, such as
+/// running actors and periodic notifications. If it gets dropped, all
+/// tasks are cancelled.
+///
+/// Each task is associated with a key, so that it can be removed from
+/// the internal `HashMap` whenever consumers want to stop tracking
+/// the corresponding task. By removing it from the `HashMap` we free
+/// up space.
+pub struct TaskMap<K>(HashMap<K, RemoteHandle<()>>);
+
+impl<K> TaskMap<K>
+where
+    K: Eq + hash::Hash,
+{
+    /// Spawn the task on the runtime and remember the handle.
+    ///
+    /// The task will be stopped if this instance of [`TaskMap`] goes
+    /// out of scope.
+    pub fn add(&mut self, key: K, f: impl Future<Output = ()> + Send + 'static) {
+        let handle = f.spawn_with_handle();
+        self.0.insert(key, handle);
+    }
+
+    /// Spawn a fallible task on the runtime and remember the handle.
+    ///
+    /// The task will be stopped if this instance of [`TaskMap`] goes
+    /// out of scope. If the task fails, the `err_handler` will be
+    /// invoked.
+    pub fn add_fallible<E, EF>(
+        &mut self,
+        key: K,
+        f: impl Future<Output = Result<(), E>> + Send + 'static,
+        err_handler: impl FnOnce(E) -> EF + Send + 'static,
+    ) where
+        E: Send + 'static,
+        EF: Future<Output = ()> + Send + 'static,
+    {
+        let fut = async move {
+            match f.await {
+                Ok(()) => {}
+                Err(err) => err_handler(err).await,
+            }
+        };
+
+        let handle = fut.spawn_with_handle();
+        self.0.insert(key, handle);
+    }
+
+    /// Remove the handle of a task from the internal `HashMap`.
+    ///
+    /// After removing a task which is present in the `HashMap`, the
+    /// `RemoteHandle` will be dropped, effectively ending the task if
+    /// it was still ongoing.
+    pub fn remove(&mut self, key: &K) {
+        self.0.remove(key);
+        self.0.shrink_to_fit(); // reap the benefits of removing the handle
```

Comment:
this line will cause us to re-allocating all the time, moving the memory to a new size that's just the size, then requesting an alloc straight away when get a new task, altogether leading to performance implications that are not worth the call in the first place. I think we could do without it - the size should not grow without end, with periodic tasks we should stabilise with N tasks per taker, rendering these constant re-allocations useless.

Reply:
Makes sense, we don't need to resize it constantly.

We could end up in a situation where we end up adding more often than we are removing, meaning that the capacity of the `HashMap` will keep growing over time. We could call `shrink_to_fit` periodically, but perhaps it's not worth it.

Reply:
I've removed the line with the last force-push.